// bandmate/src/app/shared/ui/confirm-dialog/confirm-dialog.ts
import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogModule } from '@angular/material/dialog';
import { MatButtonModule } from '@angular/material/button';

export type ConfirmDialogData = {
  title: string;
  message: string;
  confirmText?: string;
  cancelText?: string;
  tone?: 'danger' | 'neutral';
};

@Component({
  standalone: true,
  imports: [MatDialogModule, MatButtonModule],
  template: `
    <h2 mat-dialog-title>{{ data.title }}</h2>

    <div mat-dialog-content class="mt-1">
      <p class="m-0">{{ data.message }}</p>
    </div>

    <div mat-dialog-actions align="end" class="mt-3">
      <button mat-button mat-dialog-close>
        {{ data.cancelText ?? 'Cancel' }}
      </button>

      <button
        mat-raised-button
        [color]="data.tone === 'danger' ? 'warn' : 'primary'"
        [mat-dialog-close]="true"
        cdkFocusInitial
      >
        {{ data.confirmText ?? 'Confirm' }}
      </button>
    </div>
  `,
})
export class ConfirmDialogComponent {
  constructor(@Inject(MAT_DIALOG_DATA) public data: ConfirmDialogData) {}
}
